README.md
# poly-decomp-es

This is a maintained fork of poly-decomp.js, originally created by Stefan Hedman @schteppe.

`poly-decomp-es` is a library for decomposing a 2D polygon into convex pieces.

```
npm install poly-decomp-es
yarn add poly-decomp-es
```

![Decomposing a convcave polygon into convex regions](./readme.gif)

[**Launch the demo!**](https://poly-decomp-es.vercel.app/)

The library implements two algorithms, one optimal (but slow) and one less optimal (but fast).

It's is a manual port of the C++ library [Poly Decomp](https://mpen.ca/406/bayazit) by [Mark Penner](https://mpen.ca).

### Basic usage

```js
import { decomp, makeCCW, quickDecomp } from 'poly-decomp-es'

// Create a concave polygon
const concavePolygon = [
  [-1, 1],
  [-1, 0],
  [1, 0],
  [1, 1],
  [0.5, 0.5],
]

// Make sure the polygon has counter-clockwise winding. Skip this step if you know it's already counter-clockwise.
makeCCW(concavePolygon)

// Decompose into convex polygons, using the faster algorithm
const convexPolygons = quickDecomp(concavePolygon)

// ==> [  [[1,0],[1,1],[0.5,0.5]],  [[0.5,0.5],[-1,1],[-1,0],[1,0]]  ]

// Decompose using the slow (but optimal) algorithm
const optimalConvexPolygons = decomp(concavePolygon)

// ==> [  [[-1,1],[-1,0],[1,0],[0.5,0.5]],  [[1,0],[1,1],[0.5,0.5]]  ]
```

### Advanced usage

```js
import { isSimple, makeCCW, quickDecomp } from 'poly-decomp-es'

// Get user input as an array of points.
const polygon = getUserInput()

// Check if the polygon self-intersects
if (isSimple(polygon)) {
  // Reverse the polygon to make sure it uses counter-clockwise winding
  makeCCW(polygon)

  // Decompose into convex pieces
  const convexPolygons = quickDecomp(polygon)

  // Draw each point on an HTML5 Canvas context
  for (let i = 0; i < convexPolygons.length; i++) {
    const convexPolygon = convexPolygons[i]

    ctx.beginPath()
    const firstPoint = convexPolygon[0]
    ctx.moveTo(firstPoint[0], firstPoint[1])

    for (let j = 1; j < convexPolygon.length; j++) {
      const point = convexPolygon[j]
      const x = point[0]
      const y = point[1]
      c.lineTo(x, y)
    }
    ctx.closePath()
    ctx.fill()
  }
}
```

### Documentation

#### type Point = [number, number]

#### type Polygon = Point[]

#### quickDecomp(polygon: Polygon): Polygon[]

```js
import { quickDecomp } from 'poly-decomp-es'

const convexPolygons = quickDecomp(polygon)
```

Slices the polygon into convex sub-polygons, using a fast algorithm. Note that the input points objects will be re-used in the result array.

If the polygon is not simple, the decomposition will produce unexpected results.

#### decomp(polygon: Polygon): Polygon[] | false

```js
import { decomp } from 'poly-decomp-es'

const convexPolygons = decomp(polygon)
```

Decomposes the polygon into one or more convex sub-polygons using an optimal algorithm. Note that the input points objects will be re-used in the result array.

Returns false if the decomposition fails.

#### isSimple(polygon: Polygon): boolean

```js
import { isSimple, quickDecomp } from 'poly-decomp-es'

if (isSimple(polygon)) {
  // Polygon does not self-intersect - it's safe to decompose.
  const convexPolygons = quickDecomp(polygon)
}
```

Returns true if the polygon does not self-intersect. Use this to check if the input polygon is OK to decompose.

#### makeCCW(polygon: Polygon): void

```js
import { makeCCW } from 'poly-decomp-es'

console.log('Polygon with clockwise winding:', polygon)
makeCCW(polygon)
console.log('Polygon with counter-clockwise winding:', polygon)
```

Reverses the polygon, if its vertices are not ordered counter-clockwise. Note that the input polygon array will be modified in place.

#### removeCollinearPoints(polygon: Polygon, thresholdAngle = 0): void

```js
import { removeCollinearPoints } from 'poly-decomp-es'

const before = polygon.length
removeCollinearPoints(polygon, 0.1)
const numRemoved = before - polygon.length
console.log(numRemoved + ' collinear points could be removed')
```

Removes collinear points in the polygon. This means that if three points are placed along the same line, the middle one will be removed. The `thresholdAngle` is measured in radians and determines whether the points are collinear or not. Note that the input array will be modified in place.

#### removeDuplicatePoints(polygon: Polygon, precision = 0): void

```js
import { removeDuplicatePoints } from 'poly-decomp-es'

const polygon = [
  [0, 0],
  [1, 1],
  [2, 2],
  [0, 0],
]
removeDuplicatePoints(polygon, 0.01)

// polygon is now [[1,1],[2,2],[0,0]]
```


CHANGELOG.md
### Changelog

## 0.4.2

### Patch Changes

- c476da2: chore: bump various dev dependencies

##### 0.4.1

- Fix typescript declaration output

##### 0.4.0

- Use rollup to output cjs and esm bundles
- Convert to typescript

##### 0.3.0

- Added `removeDuplicatePoints`.
- `makeCCW` now returns true if the polygon was changed.
- Fixed case 5 mentioned [here](https://mpen.ca/406/bayazit) and discussed [here](https://github.com/schteppe/poly-decomp.js/issues/8).

##### 0.2.1

- Fixed bug in the collinear point removal, after this fix the algorithm is more agressive and more correct.

##### 0.2.0

- Rewrote the class based API to a minimal array-based one. See docs.

##### 0.1

- Added method `Polygon.prototype.removeCollinearPoints`.
- Added optional parameter `thresholdAngle` to `Point.collinear(a,b,c,thresholdAngle)`.


src/index.spec.ts
import { describe, expect, test } from '@jest/globals'
import type { Polygon } from './index'
import { decomp, isSimple, quickDecomp, removeDuplicatePoints, makeCCW } from './index'

const concave: Polygon = [
  [-1, 1],
  [-1, 0],
  [1, 0],
  [1, 1],
  [0.5, 0.5],
]

const circle: Polygon = []
const N = 10
for (let i = 0; i < N; i++) {
  const angle = ((2 * Math.PI) / N) * i
  circle.push([Math.cos(angle), Math.sin(angle)])
}

describe('poly-decomp-es', () => {
  test('decomp', () => {
    const circleConvexes = decomp(circle)
    expect(circleConvexes).toHaveLength(1)

    const concaveConvexes = decomp(concave)
    expect(concaveConvexes).toHaveLength(2)
  })

  test('isSimple', () => {
    const notSimple: Polygon = [
      [-1, -1],
      [0, 0],
      [1, 1],
      [0, 2],
      [-1, 1],
      [0, 0],
      [1, -1],
    ]

    expect(isSimple(concave)).toBeTruthy()
    expect(isSimple(circle)).toBeTruthy()
    expect(isSimple(notSimple)).toBeFalsy()
  })

  test('quickDecomp', () => {
    const result = quickDecomp(circle)
    expect(result).toHaveLength(1)

    const convexResult = quickDecomp(concave)
    expect(convexResult).toHaveLength(2)
  })

  test('removeDuplicatePoints', () => {
    const data: Polygon = [
      [0, 0],
      [1, 1],
      [2, 2],
      [0, 0],
    ]
    removeDuplicatePoints(data)
    expect(data).toHaveLength(3)

    const data2: Polygon = [
      [0, 0],
      [1, 1],
      [2, 2],
      [1, 1],
      [0, 0],
      [2, 2],
    ]
    removeDuplicatePoints(data2)
    expect(data2).toHaveLength(3)
  })

  test('quickDecompExtraVisibilityTestFix', () => {
    // This test checks that this bug is fixed: https://github.com/schteppe/poly-decomp.js/issues/8
    let path: Polygon = [
      [0, -134],
      [50, -139],
      [60, -215],
      [70, -6],
      [80, -236],
      [110, -120],
      [110, 0],
      [0, 0],
    ].map((point) => [2 * point[0] + 100, 1 * point[1] + 500])

    makeCCW(path)
    let polys = quickDecomp(path)
    expect(polys).toHaveLength(3)

    path = [
      [0, -134],
      [50, -139],
      [60, -215],
      [70, -6],
      [80, -236],
      [110, -120],
      [110, 0],
      [0, 0],
    ].map((point) => [3 * point[0] + 100, 1 * point[1] + 500])

    makeCCW(path)
    polys = quickDecomp(path)
    expect(polys).toHaveLength(3)

    path = [
      [0, -134],
      [50, -139],
      [60, -215],
      [70, -6],
      [80, -236],
      [110, -120],
      [110, 0],
      [0, 0],
    ].map((point) => [-3 * point[0], -point[1]])
    makeCCW(path)
    polys = quickDecomp(path)
    expect(polys).toHaveLength(3)

    path = [
      [331, 384],
      [285, 361],
      [238, 386],
      [283, 408],
      [191, 469],
      [213, 372],
      [298, 314],
      [342, 340],
    ]
    makeCCW(path)
    polys = quickDecomp(path)
    expect(polys).toHaveLength(3)
  })
})


src/poly-decomp.spec.ts
import { describe, expect, test } from '@jest/globals'
import type { Polygon } from './types'
import { decomp, isSimple, quickDecomp, removeDuplicatePoints, makeCCW } from './poly-decomp'

const concave: Polygon = [
  [-1, 1],
  [-1, 0],
  [1, 0],
  [1, 1],
  [0.5, 0.5],
]

const circle: Polygon = []
const N = 10
for (let i = 0; i < N; i++) {
  const angle = ((2 * Math.PI) / N) * i
  circle.push([Math.cos(angle), Math.sin(angle)])
}

describe('poly-decomp', () => {
  test('decomp', () => {
    const circleConvexes = decomp(circle)
    expect(circleConvexes).toHaveLength(1)

    const concaveConvexes = decomp(concave)
    expect(concaveConvexes).toHaveLength(2)
  })

  test('isSimple', () => {
    const notSimple: Polygon = [
      [-1, -1],
      [0, 0],
      [1, 1],
      [0, 2],
      [-1, 1],
      [0, 0],
      [1, -1],
    ]

    expect(isSimple(concave)).toBeTruthy()
    expect(isSimple(circle)).toBeTruthy()
    expect(isSimple(notSimple)).toBeFalsy()
  })

  test('quickDecomp', () => {
    const result = quickDecomp(circle)
    expect(result).toHaveLength(1)

    const convexResult = quickDecomp(concave)
    expect(convexResult).toHaveLength(2)
  })

  test('removeDuplicatePoints', () => {
    const data: Polygon = [
      [0, 0],
      [1, 1],
      [2, 2],
      [0, 0],
    ]
    removeDuplicatePoints(data)
    expect(data).toHaveLength(3)

    const data2: Polygon = [
      [0, 0],
      [1, 1],
      [2, 2],
      [1, 1],
      [0, 0],
      [2, 2],
    ]
    removeDuplicatePoints(data2)
    expect(data2).toHaveLength(3)
  })

  test('quickDecompExtraVisibilityTestFix', () => {
    // This test checks that this bug is fixed: https://github.com/schteppe/poly-decomp.js/issues/8
    let path: Polygon = [
      [0, -134],
      [50, -139],
      [60, -215],
      [70, -6],
      [80, -236],
      [110, -120],
      [110, 0],
      [0, 0],
    ].map((point) => [2 * point[0] + 100, 1 * point[1] + 500])

    makeCCW(path)
    let polys = quickDecomp(path)
    expect(polys).toHaveLength(3)

    path = [
      [0, -134],
      [50, -139],
      [60, -215],
      [70, -6],
      [80, -236],
      [110, -120],
      [110, 0],
      [0, 0],
    ].map((point) => [3 * point[0] + 100, 1 * point[1] + 500])

    makeCCW(path)
    polys = quickDecomp(path)
    expect(polys).toHaveLength(3)

    path = [
      [0, -134],
      [50, -139],
      [60, -215],
      [70, -6],
      [80, -236],
      [110, -120],
      [110, 0],
      [0, 0],
    ].map((point) => [-3 * point[0], -point[1]])
    makeCCW(path)
    polys = quickDecomp(path)
    expect(polys).toHaveLength(3)

    path = [
      [331, 384],
      [285, 361],
      [238, 386],
      [283, 408],
      [191, 469],
      [213, 372],
      [298, 314],
      [342, 340],
    ]
    makeCCW(path)
    polys = quickDecomp(path)
    expect(polys).toHaveLength(3)
  })
})


src/poly-decomp-es.spec.ts
import { describe, expect, test } from '@jest/globals'
import type { Polygon } from './poly-decomp-es'
import { decomp, isSimple, quickDecomp, removeDuplicatePoints, makeCCW } from './poly-decomp-es'

const concave: Polygon = [
  [-1, 1],
  [-1, 0],
  [1, 0],
  [1, 1],
  [0.5, 0.5],
]

const circle: Polygon = []
const N = 10
for (let i = 0; i < N; i++) {
  const angle = ((2 * Math.PI) / N) * i
  circle.push([Math.cos(angle), Math.sin(angle)])
}

describe('poly-decomp-es', () => {
  test('decomp', () => {
    const circleConvexes = decomp(circle)
    expect(circleConvexes).toHaveLength(1)

    const concaveConvexes = decomp(concave)
    expect(concaveConvexes).toHaveLength(2)
  })

  test('isSimple', () => {
    const notSimple: Polygon = [
      [-1, -1],
      [0, 0],
      [1, 1],
      [0, 2],
      [-1, 1],
      [0, 0],
      [1, -1],
    ]

    expect(isSimple(concave)).toBeTruthy()
    expect(isSimple(circle)).toBeTruthy()
    expect(isSimple(notSimple)).toBeFalsy()
  })

  test('quickDecomp', () => {
    const result = quickDecomp(circle)
    expect(result).toHaveLength(1)

    const convexResult = quickDecomp(concave)
    expect(convexResult).toHaveLength(2)
  })

  test('removeDuplicatePoints', () => {
    const data: Polygon = [
      [0, 0],
      [1, 1],
      [2, 2],
      [0, 0],
    ]
    removeDuplicatePoints(data)
    expect(data).toHaveLength(3)

    const data2: Polygon = [
      [0, 0],
      [1, 1],
      [2, 2],
      [1, 1],
      [0, 0],
      [2, 2],
    ]
    removeDuplicatePoints(data2)
    expect(data2).toHaveLength(3)
  })

  test('quickDecompExtraVisibilityTestFix', () => {
    // This test checks that this bug is fixed: https://github.com/schteppe/poly-decomp.js/issues/8
    let path: Polygon = [
      [0, -134],
      [50, -139],
      [60, -215],
      [70, -6],
      [80, -236],
      [110, -120],
      [110, 0],
      [0, 0],
    ].map((point) => [2 * point[0] + 100, 1 * point[1] + 500])

    makeCCW(path)
    let polys = quickDecomp(path)
    expect(polys).toHaveLength(3)

    path = [
      [0, -134],
      [50, -139],
      [60, -215],
      [70, -6],
      [80, -236],
      [110, -120],
      [110, 0],
      [0, 0],
    ].map((point) => [3 * point[0] + 100, 1 * point[1] + 500])

    makeCCW(path)
    polys = quickDecomp(path)
    expect(polys).toHaveLength(3)

    path = [
      [0, -134],
      [50, -139],
      [60, -215],
      [70, -6],
      [80, -236],
      [110, -120],
      [110, 0],
      [0, 0],
    ].map((point) => [-3 * point[0], -point[1]])
    makeCCW(path)
    polys = quickDecomp(path)
    expect(polys).toHaveLength(3)

    path = [
      [331, 384],
      [285, 361],
      [238, 386],
      [283, 408],
      [191, 469],
      [213, 372],
      [298, 314],
      [342, 340],
    ]
    makeCCW(path)
    polys = quickDecomp(path)
    expect(polys).toHaveLength(3)
  })
})


rollup.config.js
import babel from '@rollup/plugin-babel'
import resolve from '@rollup/plugin-node-resolve'
import json from '@rollup/plugin-json'
import replace from '@rollup/plugin-replace'
import filesize from 'rollup-plugin-filesize'

const extensions = ['.ts']

const babelOptions = {
  babelrc: false,
  extensions,
  exclude: '**/node_modules/**',
  babelHelpers: 'bundled',
  presets: [
    [
      '@babel/preset-env',
      {
        loose: true,
        modules: false,
        targets: '>1%, not dead, not ie 11, not op_mini all',
      },
    ],
    '@babel/preset-typescript',
  ],
}

export default [
  {
    input: `./src/poly-decomp-es.ts`,
    output: { file: `dist/poly-decomp-es.js`, format: 'esm' },
    plugins: [json(), resolve({ extensions }), babel(babelOptions), filesize()],
  },
  {
    input: `./src/poly-decomp-es.ts`,
    output: { file: `dist/poly-decomp-es.cjs.js`, format: 'cjs' },
    plugins: [
      json(),
      resolve({ extensions }),
      babel(babelOptions),
      replace({
        // Use node built-in performance.now in commonjs environments
        'globalThis.performance': `require('perf_hooks') && require('perf_hooks').performance`,
      }),
      filesize(),
    ],
  },
]


example/src/main.js
import * as decomp from 'poly-decomp-es'
import dat from 'dat.gui'

import './style.css'

const demoPolygons = [
  {
    name: 'BayazitCase1',
    data: [
      [192, 40],
      [32, 48],
      [78, 154],
      [118, 108],
      [160, 150],
    ],
  },
  {
    name: 'BayazitCase2',
    data: [
      [125, 52],
      [161, 104],
      [202, 52],
      [268, 90],
      [260, 188],
      [202, 228],
      [165, 188],
      [122, 231],
      [64, 192],
      [52, 102],
    ],
  },
  {
    name: 'BayazitCase3',
    data: [
      [160, 56],
      [177, 84],
      [196, 61],
      [214, 105],
      [316, 56],
      [318, 257],
      [219, 255],
      [202, 231],
      [184, 253],
      [166, 200],
      [53, 252],
      [56, 78],
    ],
  },
  {
    name: 'BayazitCase4',
    data: [
      [186, 56],
      [211, 110],
      [230, 55],
      [263, 53],
      [275, 84],
      [323, 98],
      [319, 246],
      [275, 245],
      [240, 213],
      [202, 243],
      [169, 218],
      [119, 248],
      [54, 246],
      [53, 81],
    ],
  },
  {
    name: 'BayazitCase5',
    data: [
      [174, 160],
      [131, 112],
      [201, 270],
      [128, 240],
      [52, 270],
      [100, 52],
    ],
  },
  {
    name: 'Simple',
    data: [
      [100, 180],
      [100, 80],
      [300, 80],
      [300, 180],
      [250, 130],
    ],
  },
  {
    name: 'Peaks',
    data: [
      [50, 220.55491449129886],
      [60, 200.9165010426762],
      [70, 171.70035249020682],
      [80, 114.90970199382247],
      [90, 165.9804978067106],
      [100, 160.4772842673997],
      [110, 84.41974678190957],
      [120, 293.8773909405506],
      [130, 63.92280957183573],
      [140, 73.75740089300058],
      [150, 61.19584194167069],
      [160, 159.87092544067778],
      [170, 247.10194630714335],
      [180, 183.8388841500984],
      [190, 117.18720369155437],
      [200, 242.94006022134025],
      [210, 206.9854959619504],
      [220, 65.33570549558465],
      [230, 263.57031109249294],
      [240, 136.10471065015298],
      [250, 300],
      [50, 300],
    ],
  },
  {
    name: 'Bayazit1',
    data: [
      [154, 124],
      [201, 475],
      [351, 359],
      [274, 505],
      [568, 485],
      [481, 342],
      [611, 453],
      [571, 130],
      [463, 240],
      [557, 86],
      [227, 98],
      [321, 265],
    ],
  },
  {
    name: 'Bayazit3',
    data: [
      [339, 279],
      [344, 142],
      [419, 316],
      [599, 112],
      [532, 351],
      [778, 290],
      [564, 422],
      [728, 615],
      [449, 513],
      [519, 720],
      [404, 594],
      [327, 717],
      [350, 495],
      [76, 580],
      [305, 436],
      [98, 338],
      [320, 370],
      [212, 217],
    ],
  },
  {
    name: 'Bayazit4',
    data: [
      [201, 291],
      [304, 477],
      [420, 274],
      [494, 464],
      [617, 259],
      [758, 458],
      [635, 66],
      [513, 324],
      [410, 97],
      [320, 327],
      [180, 119],
      [83, 369],
    ],
  },
]

let path = []
let polys = []
let reflexVertices = []
let steinerPoints = []
let mousedown = false
let colors = ['#f99', '#9f9', '#99f', '#ff9']

const canvas = document.createElement('canvas')
const pixelRatio = window.devicePixelRatio || 1
canvas.width = window.innerWidth * pixelRatio
canvas.height = window.innerHeight * pixelRatio
document.body.append(canvas)

function render() {
  const c = canvas.getContext('2d')

  // clear
  c.clearRect(0, 0, canvas.width, canvas.height)

  c.fillStyle = 'red'
  c.strokeStyle = 'black'

  if (mousedown) {
    c.beginPath()
    for (let i = 0; i < path.length - 1; i++) {
      const p = path[i]
      c.moveTo(p[0], p[1])
      c.arc(p[0], p[1], 3 * pixelRatio, 0, 2 * Math.PI)
    }
    c.fill()
  }

  // Draw convex polygons
  for (let i = 0; i < polys.length; i++) {
    c.fillStyle = colors[i % colors.length]
    const poly = polys[i]
    c.beginPath()
    let p = poly[0]
    c.moveTo(p[0], p[1])
    for (let j = 0; j < poly.length; j++) {
      p = poly[j]
      c.lineTo(p[0], p[1])
    }
    c.closePath()
    c.fill()
    c.stroke()
  }

  // Draw points
  c.fillStyle = 'black'
  for (let i = 0; i < polys.length; i++) {
    const poly = polys[i]
    c.beginPath()
    for (let j = 0; j < poly.length; j++) {
      const p = poly[j]
      c.moveTo(p[0], p[1])
      c.arc(p[0], p[1], 2 * pixelRatio, 0, 2 * Math.PI)
    }
    c.fill()
  }

  // draw the current path
  c.strokeStyle = decomp.isSimple(path) ? 'black' : 'red'
  c.lineWidth = 2 * pixelRatio
  c.beginPath()
  for (let i = 0; i < path.length; i++) {
    c.moveTo(path[i][0], path[i][1])
    c.lineTo(path[(i + 1) % path.length][0], path[(i + 1) % path.length][1])
  }
  c.fill()
  c.stroke()
  c.lineWidth = 0.5 * pixelRatio

  if (!mousedown) {
    // Draw reflex vertices
    c.fillStyle = 'blue'
    c.beginPath()
    for (let j = 0; j < reflexVertices.length; j++) {
      const p = reflexVertices[j]
      c.moveTo(p[0], p[1])
      c.arc(p[0], p[1], 4 * pixelRatio, 0, 2 * Math.PI)
    }
    c.fill()

    // Draw steiner points
    c.fillStyle = 'red'
    c.beginPath()
    for (let j = 0; j < steinerPoints.length; j++) {
      const p = steinerPoints[j]
      c.moveTo(p[0], p[1])
      c.arc(p[0], p[1], 4 * pixelRatio, 0, 2 * Math.PI)
    }
    c.fill()

    // draw point indices
    c.font = Math.floor(10 * pixelRatio) + 'px Arial'
    c.fillStyle = 'black'
    for (let i = 0; i < path.length; i++) {
      const p = path[i]
      c.fillText(i, p[0] + 3 * pixelRatio, p[1] + 3 * pixelRatio)
    }
  }
}
render()

// Setup GUI
const settings = {
  minEdgeLength: 50,
  quickDecomp: true,
  removeCollinear: true,
  collinearThreshold: 0.001,
}
const gui = new dat.GUI()

const settingsFolder = gui.addFolder('Settings')
settingsFolder.open()
settingsFolder.add(settings, 'minEdgeLength', 1, 100)
settingsFolder.add(settings, 'quickDecomp')
settingsFolder.add(settings, 'removeCollinear')
settingsFolder.add(settings, 'collinearThreshold', 0, Math.PI / 4)

const pixelRatioScale = function (p) {
  return [p[0] * pixelRatio, p[1] * pixelRatio]
}
let currentDemoPolygon = 0
const polygonsFolder = gui.addFolder('Polygons')
polygonsFolder.open()
for (let i = 0; i < demoPolygons.length; i++) {
  const demoPoly = demoPolygons[i]
  settings[demoPoly.name] = (function (p) {
    return function () {
      setPath(p.data.map(pixelRatioScale))
    }
  })(demoPoly)
  polygonsFolder.add(settings, demoPoly.name)
}
let fallback = true
const prefix = '#path='
if (window.location.hash.indexOf(prefix) === 0) {
  try {
    setPath(
      window.location.hash
        .substr(prefix.length)
        .split('/')
        .map(function (p) {
          return p.split(',').map(function (strNum) {
            return parseFloat(strNum)
          })
        })
    )
    fallback = false
  } catch (err) {}
}
if (fallback) {
  setPath(demoPolygons[0].data.map(pixelRatioScale))
}
settings.next = function () {
  currentDemoPolygon = (currentDemoPolygon + 1) % demoPolygons.length
  setPath(demoPolygons[currentDemoPolygon].data.map(pixelRatioScale))
}
polygonsFolder.add(settings, 'next')
function getMousePos(e) {
  const offset = e.target.getBoundingClientRect()
  return [e.clientX * pixelRatio - offset.left * pixelRatio, e.clientY * pixelRatio - offset.top * pixelRatio]
}

function sqdist(a, b) {
  const dx = b[0] - a[0]
  const dy = b[1] - a[1]
  return dx * dx + dy * dy
}

canvas.addEventListener('mousedown', function (e) {
  path.length = 0
  polys.length = 0
  reflexVertices.length = steinerPoints.length = 0
  path.push(getMousePos(e))
  mousedown = true
  render()
})
canvas.addEventListener('mousemove', function (e) {
  if (mousedown) {
    const point = getMousePos(e)
    const lengthSquared = sqdist(point, path[path.length - 1])
    const minLengthSquared = settings.minEdgeLength * settings.minEdgeLength
    if (lengthSquared > minLengthSquared) {
      path.push(point)
      render()
    }
  }
})

function copyPath(p) {
  return JSON.parse(JSON.stringify(p))
}

function setPath(p) {
  mousedown = false
  polys.length = reflexVertices.length = steinerPoints.length = 0
  path = copyPath(p)
  if (p.length > 3) {
    if (decomp.isSimple(path)) {
      if (decomp.makeCCW(path)) {
        console.log('The polygon is not counter-clockwise. Reversing it...')
      }

      if (settings.removeCollinear) {
        console.log(
          'Removing collinear points... ' +
            decomp.removeCollinearPoints(path, settings.collinearThreshold) +
            ' points removed.'
        )
      }
      console.log(
        'setPath([' +
          path
            .map(function (point, i) {
              return '[' + point.toString() + '/*' + i + '*/]'
            })
            .join(',') +
          '])'
      )

      console.log('Decomposing polygon of size ' + path.length + '...')
      if (settings.quickDecomp) {
        decomp.quickDecomp(path, polys, reflexVertices, steinerPoints, 25, 100, 0, path)
      } else polys = decomp.decomp(path)
      console.log('Got ' + polys.length + ' convex polygons. Done.')
    } else {
      console.log('The polygon was not simple. Aborting...')
    }
  }
  render()

  if (history.pushState) {
    const pathString = path
      .map(function (p) {
        return p[0] + ',' + p[1]
      })
      .join('/')
    history.pushState(null, null, '#path=' + pathString)
  }
}

canvas.addEventListener('mouseup', function (e) {
  setPath(path)
})
